### ISSUE TITLE
Question about the Implementation 

### ISSUE THREAD

**Issue by hjq133** [davda54/sam#17]
hi!
thanks for your awesome work!

After comparing your code with google's official implementation,
I think there may be something wrong with your code about Batch Normalization's running mean and running var.

As there are 2 forward passes when using SAM, so the running mean and running var will be computed twice on the same data input in your code. But in google's implementation, they save the **_state_**(Parameters that will not be calculated gradient， including the running mean and running var) after the first forward pass as **_inner_state_** , and then return it as the final state in this step. that's to say they only calculated the running mean and running var one time.

here is google's code : 
```python
  def get_sam_gradient(model: flax.nn.Model, rho: float):
    """Returns the gradient of the SAM loss loss, updated state and logits.
    See https://arxiv.org/abs/2010.01412 for more details.
    Args:
      model: The model that we are training.
      rho: Size of the perturbation.
    """
    # compute gradient on the whole batch
    (_, (inner_state, _)), grad = jax.value_and_grad(
        lambda m: forward_and_loss(m, true_gradient=True), has_aux=True)(model)
    if FLAGS.sync_perturbations:
      if FLAGS.inner_group_size is None:
        grad = jax.lax.pmean(grad, 'batch')
      else:
        grad = jax.lax.pmean(
            grad, 'batch',
            axis_index_groups=local_replica_groups(FLAGS.inner_group_size))
    grad = dual_vector(grad)
    noised_model = jax.tree_multimap(lambda a, b: a + rho * b,
                                     model, grad)
    (_, (_, logits)), grad = jax.value_and_grad(
        forward_and_loss, has_aux=True)(noised_model)
    return (inner_state, logits), grad
```

so, the little difference may affect the performance when evaluation, right?

**Comment by davda54** [comment]
Hi, thank you for catching this hidden issue :) You are absolutely right that it can negatively influence the performance because the running statistics will be biased.

However, I'm not sure if I should clutter the example to bypass this issue (right now, I can't think of any clean solution — other than explicitly turning on/off the running averages before entering each batch norm). Have you tested the actual impact of this bug?

**Comment by hjq133** [comment]
I didn't think of a simpler solution, too. Yesterday I used resnet-50 to test on imagenet. This bug has little effect on the final result. Maybe it can be ignored.

**Comment by evanatyourservice** [comment]
I thought about this too. Can this be solved by dividing BN's momentum by 2?

**Comment by davda54** [comment]
Actually, the easiest solution is to call `model.eval()` before the second forward pass. However, the catch is that it will also turn off all dropout layers, not only the running averages. Dividing the momentum by 2 won't work, the problem is that the running averages are computed from both passes, which does not necessarily mean they are larger.

**Comment by evanatyourservice** [comment]
Oh ok so we could do
```
for module in net.modules():
    if isinstance(module, BatchNorm):
        module.eval()
```
then call net.train() after the forward and backward pass to put it back. We should probably turn them off during the second pass, right?

**Comment by davda54** [comment]
Yes, that sounds reasonable, thanks :)

**Comment by slala2121** [comment]
To confirm so the modified code would look like this?

`from sam import SAM
...

model = YourModel()
base_optimizer = torch.optim.SGD  # define an optimizer for the "sharpness-aware" update
optimizer = SAM(model.parameters(), base_optimizer, lr=0.1, momentum=0.9)
...


def disable_bn(net):
  for module in net.modules():
      if isinstance(module, BatchNorm):
          module.eval()
def enable_bn(net):
   net.train()


for input, output in data:
  def closure():
    loss = loss_function(output, model(input))
    loss.backward()
    return loss

  loss = loss_function(output, model(input))
  loss.backward()
  disable_bn(model)
  optimizer.step(closure)
  optimizer.zero_grad()
  enable_bn(model)
...`

**Comment by evanatyourservice** [comment]
That looks right. I wonder what the true downside is to leaving it in train mode for both passes, if it would actually hurt at all... in the original jax implementation it looks like they just let it calculate every pass, so it might not be worth the overhead.

**Comment by guozhiyao** [comment]
The first forward pass use the mean and variance of input, if we set the model to be eval mode in the second forward pass, the mean and variance will be the running mean and running variance. Would it be a problem?

**Comment by davda54** [comment]
It depends :)

> Actually, the easiest solution is to call `model.eval()` before the second forward pass. However, the catch is that it will also turn off all dropout layers, not only the running averages. Dividing the momentum by 2 won't work, the problem is that the running averages are computed from both passes, which does not necessarily mean they are larger.

**Comment by erhul** [comment]
```
def disable_bn(net):
    """
    将网络中的 BatchNorm 模块设置为推理模式，但保留网络整体的训练/推理状态。
    """
    for module in net.modules():
        if isinstance(module, (nn.BatchNorm1d, nn.BatchNorm2d, nn.BatchNorm3d)):
            module.eval()  # 仅冻结运行统计量，但不冻结权重

def enable_bn(net):
    """
    将网络中的 BatchNorm 模块重新设置为训练模式。
    """
    for module in net.modules():
        if isinstance(module, (nn.BatchNorm1d, nn.BatchNorm2d, nn.BatchNorm3d)):
            module.train()  # 恢复训练状态
```
